File: 144.Binary-Tree-Preorder-Traversal/recursive.js
var preorderTraversal = function (root, answer) {
    if (answer === void 0) { answer = []; }
    if (root) {
        answer.push(root.val);
    }
    preorderTraversal(root.right, answer);
    preorderTraversal(root.left, answer);
    return answer;
};

File: todo/144.Binary-Tree-Preorder-Traversal/java/src/com/example/helloworld/Recursive.java
package com.example.helloworld;

import java.util.ArrayList;
import java.util.List;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

public class Recursive {
    public List<Integer> preorderTraversal(TreeNode root) {
        return preorderTraversal(root, new ArrayList<Integer>());
    }

    public List<Integer> preorderTraversal(TreeNode root, ArrayList<Integer> list) {
        if (root == null) {
            return list;
        }
        list.add(root.val);
        preorderTraversal(root.left, list);
        preorderTraversal(root.right, list);
        
        return list;
    }
}
